Storybook and TS migration of ReadMore component

sunyatasattva opened the pull request:
This PR adds stories for the `ReadMore` component and migrates it to TypeScript. It also migrates a bunch of associated functions and components.

Fixes #5254

### Testing

### Manual Testing

How to test the changes in this Pull Request:

1. Run `npm run storybook`
2. Explore the [ReadMore stories](http://localhost:6006/?path=/story/woocommerce-blocks-base-components-readmore--default). You should see one default story with the appropriate controls.

> ⚠️  **Please note:** Component will not properly react with control changes. This is because of the way component was written. Fixing this is not part of this PR. See #5299

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
* [x] Same as above

sunyatasattva committed:
Update packages to latest Storybook deps

sunyatasattva committed:
Migrate `ReadMore` component and its utils to TypeScript

sunyatasattva committed:
Add type definitions for `trim-html` package

sunyatasattva committed:
Migrate `ReadMore` stories to newest Storybook

sunyatasattva committed:
Add correct packages

sunyatasattva committed:
Refactor story

sunyatasattva committed:
Migrate test to TypeScript

tjcafferkey commented on assets/js/base/components/read-more/index.tsx:
```diff
@@ -38,18 +104,22 @@ class ReadMore extends Component {
 		if ( this.props.children ) {
 			const { maxLines, ellipsis } = this.props;
 
+			if ( ! this.reviewSummary.current || ! this.reviewContent.current )
```
This seems inconsistent with the syntax in the rest of this file, and codebase. Can we open this with curly braces?

tjcafferkey commented on assets/js/base/components/read-more/index.tsx:
```diff
@@ -70,7 +140,7 @@ class ReadMore extends Component {
 	}
 
-	getButton() {
+	getButton(): JSX.Element | undefined {
```
Can we return `null` instead of `undefined` so its consistent with other methods that render JSX e.g. the `render()`

sunyatasattva replied:
Mm, sure though I was just typing the existing situation here. As I'm not sure about the current test coverage, this could potentially have side effects. Though likely people will check for falsy values and not specifically for `undefined` and, even more likely, nobody outside of the component is consuming this function which should be `private`. 

I was just super cautious I guess 😬

tjcafferkey replied:
Yeah also I don't think they will be checking the explicit value of a method that renders markup but worth checking if you think so

tjcafferkey commented on assets/js/base/components/read-more/index.tsx:
```diff
@@ -38,18 +104,22 @@ class ReadMore extends Component {
+
 			const lineHeight = this.reviewSummary.current.clientHeight + 1;
 			const reviewHeight = this.reviewContent.current.clientHeight + 1;
+
```
Added space accidentally?

sunyatasattva replied:
Good catch, thanks! Yea, I had a console.log there 😬

tjcafferkey commented on assets/js/base/components/read-more/utils.ts:
```diff
@@ -20,35 +30,39 @@ export const truncateHtml = ( html, length, ellipsis = '...' ) => {
+	markers: Markers,
+	currentHeight: number,
+	maxHeight: number
 ) => {
```
Can we type the return value here as `Markers`?

tjcafferkey commented on assets/js/base/components/read-more/utils.ts:
```diff
@@ -20,35 +30,39 @@ export const truncateHtml = ( html, length, ellipsis = '...' ) => {
+	originalContent: string,
+	targetElement: HTMLElement,
+	maxHeight: number
+) => {
```
Can we type the return value here? Im assuming `number`?

tjcafferkey commented on assets/js/base/components/read-more/stories/index.tsx:
```diff
@@ -1,15 +1,24 @@
 
-export const Default = () => (
-	<ReadMore maxLines={ 2 }>
+const LongText = (
```
Can we type this as `JSX.Element`?

sunyatasattva committed:
Address PR feedback

sunyatasattva commented:
@tjcafferkey Addressed the feedback! Thanks for the pointers. I have one doubt though about the whole explicit return types, but especially about `const LongText: JSX.Element` thing: wouldn't you say that's a bit unnecessary? TypeScript already knows that and it being a `const` you can't really reassign it to anything different. What's the benefit there? (I fixed it though)

tjcafferkey commented:
>wouldn't you say that's a bit unnecessary? TypeScript already knows that

@sunyatasattva yeah you're right I think about this. Theres no need for it, my bad.

sunyatasattva committed:
Revert some PR feedback

sunyatasattva commented:
@tjcafferkey Reverted the explicit typing on the LongText. Let me know if there is anything else 🙏
